### PR TITLE
Feature/fix issue 1673

### PR DESCRIPTION
**Pull request by tingyuz**
Add framework project for security-secretstore-setup so that the features can be added/updated later. Please track the features under project #13.

### CODE REVIEW

**Comment by tsconn23** [comment]
@tingyuz 
```
10:12:57 ./bin/test-attribution-txt.sh
10:12:58 An Attribution.txt file for cmd/security-secretstore-setup is missing, please add
10:12:59 make: *** [test] Error 1
```

**Comment by tingyuz** [comment]
> @tingyuz
> 
> ```
> 10:12:57 ./bin/test-attribution-txt.sh
> 10:12:58 An Attribution.txt file for cmd/security-secretstore-setup is missing, please add
> 10:12:59 make: *** [test] Error 1
> ```

Attribution.txt added into the project.

**Comment by tingyuz** [comment]
recheck

**Comment by jdharms** [comment]
recheck

File: internal/security/secretstore/config.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	LogLevel string
+	Title    string
+}
+
+type SecretServiceInfo struct {
+	Scheme               string
+	Server               string
+	Port                 int
+	CertPath             string
+	CaFilePath           string
```
Comment (mkbhanda) [inline internal/security/secretstore/config.go]: A little baffling
<foo>path, <foo>file
but we have a CertPath and CertFilePath (is this path plus file name?, then do we need CertPath?)
CaFilePath -- is this to a key. or the CA cert?

File: internal/pkg/usage/usage.go [hunk 2]
```diff
@@ -67,6 +67,17 @@ Common Options:
 	-h, --help                      Show this message
 `
 
+var securitySecretStoreSetupUsageStr = `
+Usage: %s [options]
+Server Options:
+	--insureskipverify=true/false			Indicates if skipping the server side SSL cert verifcation, similar to -k of curl
```
Comment (mkbhanda) [inline internal/pkg/usage/usage.go]: would prefer to just call this insecure, keeping in line with curl

File: cmd/security-secretstore-setup/secretstore-init.sh [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/dumb-init /bin/sh
+set -e
+
+
+# this is the placeholder for secretstore-init based on the discussion with Bryon ( see details in issue #1673 of edgex-go)
+# the name and content may be changed later
```
Comment (mkbhanda) [inline cmd/security-secretstore-setup/secretstore-init.sh]: just in case we care, no newline at the end

File: internal/security/secretstore/config.go (hunk 1 above)
Comment (anonymouse64) [reply internal/security/secretstore/config.go]: I read it as `CertPath` is the path to look for system certs to verify the TLS host, while `CertFilePath` is a file path to a specific file that should be used to verify the TLS host.

Comment (anonymouse64) [reply internal/security/secretstore/config.go]: but agreed it's confusing and could use a rename

File: cmd/security-secretstore-setup/main.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * @author: Alain Pulluelo, ForgeRock AS
+ * @author: Tingyu Zeng, Dell
+ *
+ *******************************************************************************/
+package main
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/main.go]: @jim-wang-intel   Note this and check our pending commits for the same issue ( whitespace line between the package declaration and the copyright header )

File: cmd/security-secretstore-setup/secretstore-init.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/dumb-init /bin/sh
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: @anonymouse64 Can we declare this as docker-specific file or can you suggest a docker-specific version of the file?  I asked @tingyuz for this hook so that we can spawn the token-issuing server as a subprocess and have proper process reaping and signal handling.  cc: @jim-wang-intel 

See also https://github.com/Yelp/dumb-init/blob/master/README.md
See also Vault's boilelplate: https://github.com/hashicorp/docker-vault/tree/master/0.X

[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/security-secretstore-setup/secretstore-init.sh]: I guess before we go further with figuring out what to do with `dumb-init` here, my larger concern is about this being a shell script at all... See my comment in the actual review

[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: @anonymouse64 Yes!  In fact, in TingYu's refactoring, the token issuing code has been completely removed because the Kong setup has been changed for the time being to use the root token.  Why do you prefer a go program that forks subprocesses vs a shell script that just does (subprocess & ; subprocess &)
cc: @jim-wang-intel

File: cmd/security-secretstore-setup/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+WORKDIR /
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/res/docker/configuration.toml /res/configuration.toml
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/security-secretstore-setup .
+
+ENTRYPOINT ["secretstore-init.sh"]
+
+#ENTRYPOINT ["./security-secretstore-setup","--init=true"]
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/Dockerfile]: BUG!

`cmd/security-secretstore-setup/secretstore-init.sh` was supposed to
```sh
exec ./security-secretstore-setup --init=true
```
at the end.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 3 above)
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: BUG!  Add

```sh
exec ./security-secretstore-setup --init=true
```

to the end of this line.  Otherwise the container will just run this script script and do nothing!

File: cmd/security-secretstore-setup/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2019: Dell Technologies, Inc.'
+
+WORKDIR /
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/res/docker/configuration.toml /res/configuration.toml
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/security-secretstore-setup .
+
+ENTRYPOINT ["secretstore-init.sh"]
```
Comment (bnevis-i) [inline cmd/security-secretstore-setup/Dockerfile]: Missing

```sh
RUN apk --no-cache add dumb-init
```

otherwise secretstore-init.sh will crash.

There should also be defensive code here to `chmod 755 secretstore-init.sh` or container init could also crash.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/security-secretstore-setup/secretstore-init.sh]: While I do write quite a bit of shell code myself in snapd and in authoring snaps (see all the scripts in edgex-go for example under snap/), shell code is very often brittle and difficult to review. There are all kinds of gotchas and weird bugs with different versions of shells you use and while you solve some of that by sticking to strict POSIX with `/bin/sh`, you lose out on a whole lot of useful constructs that greatly simplify the code you write. Furthermore, the project as a whole has decided on Go as the primary language of development and as such I think we should be consistent. It will be easier for other non-security developers to review, it is more portable to different platforms (i.e. if you had a single go program you don't need to worry about all of these issues with forking and reaping and signals and such in docker containers, and you also get windows support pretty easily and don't need to worry about the differences in shell environments between Mac and Linux)

To summarize:
* Shell code is very brittle and rots very quickly
* Go code is more readable 
* Shell code is difficult to write _always_ correctly without extensive use of something like `shellcheck`
* Go has been adopted by the wider EdgeX community, the security services should follow this pattern
* Go code is more easily ported than shell code

I guess finally, my question to you would be: Why do we need the go program to fork subprocesses? Why can't the go program just do the operations it needs to? I see your proposals on the various issues about splitting up the responsibilities but what I don't see is a justification for why we should split these things up other than it separates responsibilities.

[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: @anonymouse64  Primary reason is our discussion earlier about future capabilities for the Vault token server: namely, be able to have different implementations.  One implementation would create vault tokens statically and it would be wired up into the health check for secret store readiness.  The other one we discussed was a domain socket API that could create them on the fly.  Given that go plugins are not a viable option at this time, the only other alternatives we have for that type of configuration are (1) separate processes, or (2) include both options into the same executable.

I think it important to decide now what architectural pattern we are going to use when we want to give users a choice.

[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: @anonymouse64 Other similar choices in the architecture that are coming up include how we add Vault master key protection and be able to extend it to hardware solutions.

[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/security-secretstore-setup/secretstore-init.sh]: well at the minimum we should open up discussion around switching to shell instead of go to the wider community and not just hash it out in this PR...

Also we already have a pattern for this in the SMA, by providing a configuration.toml specification for what binary to run, that could be used here too. If there are concerns about reaping those processes, I think we should look at concrete issues/bugs there before deciding that forking programs in Go is a non-starter, because AFAICT Go does it correctly...

[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: @anonymouse64 That is agreeable to me.  In that case we should switch the dockerfile back to FROM scratch and nix the entrypoint script.

File: cmd/security-secretstore-setup/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/Dockerfile]: See https://github.com/edgexfoundry/edgex-go/pull/1719/files/aa672477b08f26601702723eda8d51e78e880a30#r319581789

Recommendation is to go back to `FROM scratch` and put ENTRYPOINT back the way it was.

File: internal/pkg/usage/usage.go (hunk 2 above)
Comment (jdharms) [reply internal/pkg/usage/usage.go]: This option is read into `var ensureSkipVerify bool`, so maybe insecure was not the intention here.  @tingyuz ?

Comment (jdharms) [reply internal/pkg/usage/usage.go]: Nevermind... I think it's just typos both ways.

File: cmd/security-secretstore-setup/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/Dockerfile]: @bnevis-i I thought your proposal is to run the script at the end of the dockerfile and all the binaries will be called in the script, so I have this dummy script called at the end. So if we go back to call the binary directly, where will be script be residing?

File: cmd/security-secretstore-setup/res/docker/configuration.toml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+[Logging]
+EnableRemote = false
+File = './logs/security-secretstore-setup.log'
+
+[SecretService]
+scheme = "https"
+server = "edgex-vault"
+port = 8200
+certpath = "v1/secret/edgex/pki/tls/edgex-kong"
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/res/docker/configuration.toml]: These paths will not be deprecated, but I may be moved to different configuration files depending on the binary.

File: internal/pkg/usage/usage.go (hunk 2 above)
Comment (tingyuz) [reply internal/pkg/usage/usage.go]: this should be "insecureXXX", will change it.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 3 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/secretstore-init.sh]: I may misunderstand you @bnevis-i since I thought we need to call the script in which all the binaries will be called. Currently secretstore-init.sh will have nothing inside. When we implement security-sercretstore-setup we will just add the binary into the file.

File: cmd/security-secretstore-setup/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnevis-i) [reply cmd/security-secretstore-setup/Dockerfile]: See discussion on `cmd/security-secretstore-setup/secretstore-init.sh`

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 3 above)
Comment (bnevis-i) [reply cmd/security-secretstore-setup/secretstore-init.sh]: See discussion for line 1 of this file above.

File: cmd/security-secretstore-setup/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/Dockerfile]: as we discussed from last Friday this is not a bug. This serves as the holder project, so it will not include any implementation. Line of "exec ./security-secretstore-setup --init=true" should be added into script if we decide to adopt script, or added into the docker file if we decide not to use script. in this sense I would prefer to keep it as is now.

File: internal/security/secretstore/config.go (hunk 1 above)
Comment (tingyuz) [reply internal/security/secretstore/config.go]: I would like to keep it as is for consistence. If we decide to divide it into a couple of smaller modules, we will have chance to address them there.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 3 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/secretstore-init.sh]: As I mentioned above this is not a bug but a holder project. This line needs to be added into different place depending on the result of future discussion.

File: cmd/security-secretstore-setup/main.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+	var initNeeded bool
+	var ensureSkipVerify bool
+	var configFileLocation string
+	var waitInterval int
+	var useProfile string
+	var useRegistry bool
+
+	flag.BoolVar(&initNeeded, "init", false, "run init procedure for security service.")
+	flag.BoolVar(&ensureSkipVerify, "insureskipverify", true, "skip server side SSL verification, mainly for self-signed cert")
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: will change this as "insecureskipVerify".

File: cmd/security-secretstore-setup/main.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	flag.StringVar(&configFileLocation, "configfile", "res/configuration.toml", "configuration file")
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: will add -c option.

File: cmd/security-secretstore-setup/res/configuration.toml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+[Writable]
+LogLevel = 'DEBUG'
+
+[Logging]
+EnableRemote = false
+File = './logs/security-secretstore-setup.log'
+
+[SecretService]
+scheme = "https"
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/res/configuration.toml]: Added -p and -r options in the command line parameter.  For the fields name in the toml file, I'd like to change them after we decide if we have one binary or multiple with different configurations - these fields are not being used currently and are not parsed.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/secretstore-init.sh]: switched Dockerfile back to scratch per the discussion

File: cmd/security-secretstore-setup/Dockerfile (hunk 7 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/Dockerfile]: I have removed the .sh file and related update for Dockerfile to avoid any confusing. If we need to adopt script approach after tomorrow's WG meeting we can alwasy update Dockerfile to add script.

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/security-secretstore-setup/res/configuration.toml]: okay, in that case this change can wait for another PR

File: cmd/security-secretstore-setup/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/Dockerfile]: As we decided to remove the script option during last week's meeting, this falls back to call the binary directly.

File: internal/security/secretstore/config.go (hunk 1 above)
Comment (tingyuz) [reply internal/security/secretstore/config.go]: Create issue #1762  to track the incoming update.

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 3 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/secretstore-init.sh]: Changed back to original method of calling binary.

File: cmd/security-secretstore-setup/res/configuration.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/res/configuration.toml]: Tracked under #1762

File: cmd/security-secretstore-setup/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: Fixed.

File: cmd/security-secretstore-setup/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: Tracked under issue #1764

[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: tracked under #1764

File: cmd/security-secretstore-setup/secretstore-init.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/secretstore-init.sh]: This is no longer valid as we decide to remove script.

File: cmd/security-secretstore-setup/main.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+	flag.IntVar(&waitInterval, "wait", 30, "time to wait between checking Vault status in seconds.")
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: tracked under #1765.

[the remark this answers is not in the document]
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: tracked under #1765.

File: cmd/security-secretstore-setup/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2019: Dell Technologies, Inc.'
+
+WORKDIR /
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/res/docker/configuration.toml /res/configuration.toml
+
+COPY --from=builder /edgex-go/cmd/security-secretstore-setup/security-secretstore-setup .
+
+#ENTRYPOINT ["secretstore-init.sh"]
```
Comment (anonymouse64) [inline cmd/security-secretstore-setup/Dockerfile]: since we don't have this anymore could we delete it from the Dockerfile?

File: cmd/security-secretstore-setup/main.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	var initNeeded bool
+	var insecureSkipVerify bool
+	var configFileLocation string
+	var waitInterval int
+	var useProfile string
+	var useRegistry bool
+
+	flag.BoolVar(&initNeeded, "init", false, "run init procedure for security service.")
```
Comment (anonymouse64) [inline cmd/security-secretstore-setup/main.go]: @tingyuz can we queue up a discussion/review of the CLI interface for security-secretstore-setup like we've had for security-secrets-setup for next week? Not for this PR should be done before Fuji

File: cmd/security-secretstore-setup/main.go [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+	flag.BoolVar(&insecureSkipVerify, "insecureSkipVerify", true, "skip server side SSL verification, mainly for self-signed cert")
```
Comment (anonymouse64) [inline cmd/security-secretstore-setup/main.go]: thanks for fixing the name here, but the default for this is wrong it should be false by default so that if you want to operate in insecure mode you need to add additional CLI args

```suggestion
	flag.BoolVar(&insecureSkipVerify, "insecureSkipVerify", false, "skip server side SSL verification, mainly for self-signed cert")
```

File: cmd/security-secretstore-setup/main.go (hunk 14 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: Sure we will put it into agenda.

File: cmd/security-secretstore-setup/main.go (hunk 15 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: Fixed.

File: cmd/security-secretstore-setup/Dockerfile (hunk 13 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/Dockerfile]: Fixed.
